Mathlib/Data/Matroid/IndepAxioms.lean

```diff
@@ -0,0 +1,515 @@
+
+/-- An `M : IndepMatroid α` gives a `Matroid α` whose bases are the maximal `M`-independent sets. -/
+@[simps] protected def matroid (M : IndepMatroid α) : Matroid α where
+  E := M.E
+  Base := (· ∈ maximals (· ⊆ ·) {I | M.Indep I})
+  exists_base := ( by
+      obtain ⟨B, ⟨hB,-,-⟩, hB₁⟩ :=
+        M.indep_maximal M.E rfl.subset ∅ M.indep_empty (empty_subset _)
+      exact ⟨B, ⟨hB, fun B' hB' h' ↦ hB₁ ⟨hB', empty_subset _,M.subset_ground B' hB'⟩ h'⟩⟩)
+  base_exchange := ( by
+    rintro B B' ⟨hB, hBmax⟩ ⟨hB',hB'max⟩ e he
+    have hnotmax : B \ {e} ∉ maximals (· ⊆ ·) {I | M.Indep I}
+    { simp only [mem_maximals_setOf_iff, diff_singleton_subset_iff, not_and, not_forall,
+        exists_prop, exists_and_left]
+      exact fun _ ↦ ⟨B, hB, subset_insert _ _, by simpa using he.1⟩ }
+
+    obtain ⟨f,hf,hfB⟩ := M.indep_aug (M.indep_subset hB (diff_subset B {e})) hnotmax ⟨hB',hB'max⟩
+    simp only [mem_diff, mem_singleton_iff, not_and, not_not] at hf
+
+    have hfB' : f ∉ B := by (intro hfB; obtain rfl := hf.2 hfB; exact he.2 hf.1)
+
+    refine' ⟨f, ⟨hf.1, hfB'⟩, by_contra (fun hnot ↦ _)⟩
+    obtain ⟨x,hxB, hind⟩ :=  M.indep_aug hfB hnot ⟨hB, hBmax⟩
+    simp only [mem_diff, mem_insert_iff, mem_singleton_iff, not_or, not_and, not_not] at hxB
+    obtain rfl := hxB.2.2 hxB.1
+    rw [insert_comm, insert_diff_singleton, insert_eq_of_mem he.1] at hind
+    exact not_mem_subset (hBmax hind (subset_insert _ _)) hfB' (mem_insert _ _) )
+  maximality :=
+  ( by
+      rintro X hXE I ⟨hB, hB, hIB⟩ hIX
+      obtain ⟨J, ⟨hJ, hIJ, hJX⟩, hJmax⟩ := M.indep_maximal X hXE I (M.indep_subset hB.1 hIB) hIX
+      obtain ⟨BJ, hBJ⟩ := M.indep_maximal M.E rfl.subset J hJ (M.subset_ground J hJ)
+      refine' ⟨J, ⟨⟨BJ,_, hBJ.1.2.1⟩ ,hIJ,hJX⟩, _⟩
+      · exact ⟨hBJ.1.1, fun B' hB' hBJB' ↦ hBJ.2 ⟨hB',hBJ.1.2.1.trans hBJB',
+          M.subset_ground _ hB'⟩ hBJB'⟩
+      simp only [maximals, mem_setOf_eq, and_imp, forall_exists_index]
+      rintro A B' (hBi' : M.Indep _) - hB'' hIA hAX hJA
+      simp only [mem_setOf_eq, and_imp] at hJmax
+      exact hJmax (M.indep_subset hBi' hB'') hIA hAX hJA )
+  subset_ground := ( fun B hB ↦ M.subset_ground B hB.1 )
+
+@[simp] theorem matroid_indep_iff {M : IndepMatroid α} {I : Set α} :
+    M.matroid.Indep I ↔ M.Indep I := by
+  simp only [IndepMatroid.matroid, indep_iff_subset_base]
+  refine ⟨fun ⟨_, ⟨hB,_⟩, hIB⟩ ↦ M.indep_subset hB hIB, fun h ↦ ?_⟩
+  obtain ⟨B, hB⟩ := M.indep_maximal _ rfl.subset I h (M.subset_ground _ h)
+  simp only [mem_maximals_setOf_iff, and_imp] at *
+  exact ⟨B, ⟨hB.1.1, fun J hJ hBJ ↦
+    hB.2 hJ (hB.1.2.1.trans hBJ) (M.subset_ground _ hJ) hBJ⟩, hB.1.2.1⟩
+
+/-- An independence predicate satisfying the finite matroid axioms determines a matroid,
+  provided independence is determined by its behaviour on finite sets.
+  This fundamentally needs choice, since it can be used to prove that every vector space
+  has a basis. -/
+@[simps E] protected def ofFinitary (E : Set α) (Indep : Set α → Prop)
```

Comment:
Similarly, would it make sense to define a `FinitaryMatroid` structure and make `IndepMatroid.ofFinitary` be a function `FinitaryMatroid α -> IndepMatroid α`?

Maybe there are too many variations though and this is fine for now.

Reply:
Yes, I think it is ok as it stands.